tokio/src/sync/broadcast.rs

```diff
@@ -746,22 +794,49 @@ where
     /// }
     /// ```
     pub fn try_recv(&mut self) -> Result<T, TryRecvError> {
-        let guard = self.recv_ref()?;
+        let guard = self.recv_ref(None)?;
         guard.clone_value().ok_or(TryRecvError::Closed)
     }
 
-    #[doc(hidden)] // TODO: document
+    #[doc(hidden)]
+    #[deprecated(since = "0.2.21", note = "use async fn recv()")]
     pub fn poll_recv(&mut self, cx: &mut Context<'_>) -> Poll<Result<T, RecvError>> {
-        if let Some(value) = ok_empty(self.try_recv())? {
-            return Poll::Ready(Ok(value));
+        use Poll::{Pending, Ready};
+
+        struct Guard<'a, T> {
+            waiter: Option<Pin<Box<UnsafeCell<Waiter>>>>,
+            receiver: &'a mut Receiver<T>,
         }
 
-        self.register_waker(cx.waker());
+        impl<'a, T> Drop for Guard<'a, T> {
+            fn drop(&mut self) {
+                self.receiver.waiter = self.waiter.take();
+            }
+        }
```

Comment:
The purpose of the guard is not extremely obvious — can we add a comment explaining it?
It _appears_ that this is to prevent the `Receiver`'s drop impl from dropping the waiter, but when we drop the guard, we just put the boxed waiter back in the receiver, so the receiver will still ultimately drop it?

Reply:
I can add some comments around this, but it would be a very big memory mgmt but to drop the waiter while it is in the linked list. As `Receiver::drop` handles this case, I opted to place the waiter back on `Receiver` instead of duplicating that logic.

Reply:
What I don't understand is why the guard is necessary for this. Can't the waiter just never leave the `Receiver`?

Reply:
Borrow checker does not like this.

Reply:
ahh okay, weird. this makes sense in that case

Reply:
I added a comment to clarify.